### PR TITLE
Create Index

### PR DESCRIPTION
**Pull request by pbardea**
Addressed #13 

This PR allows users to create an index on a property and/or type.

Usage:
1) `CREATE INDEX ON :person(name)`
OR
2) `CREATE INDEX ON :(name)`

The first usage creates an index on all nodes of type `person` and will index the `name` property of that node.
The second on the other hand will create an index on vertices of all types that have the `name` property.

Note: indexes may only be created on properties of type String.

### CODE REVIEW

**Comment by sachdevs** [comment]
nice on coverage 👍

File: src/main/java/ca/waterloo/dsg/graphflow/graph/Graph.java [hunk 1]
```diff
@@ -573,6 +587,20 @@ public int getHighestPermanentVertexId() {
...
 
+    /**
+     * @param vertexId The ID of the vertex whose properties will be fetched.
+     * @return The properties of the given vertex if the vertex exists. Otherwise null.
+     */
+    private Map<Short, Object> getOldVertexProperties(int vertexId) {
+        Map<Short, Object> oldVertexProperties = null;
+        try {
+            oldVertexProperties = VertexPropertyStore.getInstance().getProperties(vertexId);
+        } catch (NoSuchElementException e) {
```
Comment (semihsalihoglu-uw) [inline src/main/java/ca/waterloo/dsg/graphflow/graph/Graph.java]: Why are you catching this? More specifically, why would this throw an exception. If we want to return null in this case, I think it's better to update VPS to return null when it does not find it. Or if for some reason there is some called that wants to throw an exception, we can have two versions of this method: getOldVertexProperties and getOldVertexPropertiesOrNull or something. But one version that returns null might be better for now.

File: src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+     * given property {@code String} value. If the given type is Type.ANY, an index will be
+     * created for all types and the index will remember to index any future types on this property.
+     *
+     * @param typeKey The type of vertices to index (possibly Type.ANY).
+     * @param propertyKey The property to index on as a {@code Short}.
+     */
+    public void createIndex(Short typeKey, Short propertyKey) {
+        indexedPropertiesAndTypes.putIfAbsent(propertyKey, new HashSet<>());
+        Set<Short> currentIndexedTypes = indexedPropertiesAndTypes.get(propertyKey);
+
```
Comment (semihsalihoglu-uw) [inline src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java]: remove blank line.

File: src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+    }
+
+    /**
+     * Return whether or not a type-property pair is indexed.
+     *
+     * @param type The type to query.
+     * @param property The property to query.
+     * @return Whether or not the type-property pair is indexed.
+     */
+    boolean isIndexed(short type, short property) {
```
Comment (semihsalihoglu-uw) [inline src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java]: I think this is visible for testing only. I'd annotate it as such.

File: src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+     *
+     * @param currentIndexedTypes The set of types that are currently indexed.
+     * @param typeKey The type that is going to be indexed.
+     * @param propertyKey The property that is going to be indexed.
+     */
+    private void validateIndex(Set<Short> currentIndexedTypes, Short typeKey, Short propertyKey) {
+        if (currentIndexedTypes.contains(TypeAndPropertyKeyStore.ANY)) {
+            String propertyString = TypeAndPropertyKeyStore.getInstance().
+                mapShortPropertyKeyToString(propertyKey);
+            if (typeKey == TypeAndPropertyKeyStore.ANY) {
```
Comment (semihsalihoglu-uw) [inline src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java]: constants first.

File: src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+                    "exists on all types, so cannot create an index on a more specific type " +
+                    typeString + ".";
+                throw new IndexAlreadyExistsException(exceptionMessage);
+            }
+        }
+        if (currentIndexedTypes.contains(typeKey)) {
+            String propertyString = TypeAndPropertyKeyStore.getInstance().
+                mapShortPropertyKeyToString(propertyKey);
+            String typeString = TypeAndPropertyKeyStore.getInstance().mapShortToStringType(typeKey);
+            String exceptionMessage = "An index on property " + propertyString + " for type " +
```
Comment (semihsalihoglu-uw) [inline src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java]: You can inline the exception message.

File: src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+            } else {
+                String typeString = TypeAndPropertyKeyStore.getInstance().
+                    mapShortToStringType(typeKey);
+                String exceptionMessage = "The index on property " + propertyString + " already " +
+                    "exists on all types, so cannot create an index on a more specific type " +
+                    typeString + ".";
+                throw new IndexAlreadyExistsException(exceptionMessage);
+            }
+        }
+        if (currentIndexedTypes.contains(typeKey)) {
```
Comment (semihsalihoglu-uw) [inline src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java]: I think this should be else if, and not if. Else if of  if(currentIndexedTypes.contains(TypeAndPropertyKeyStore.ANY)) above.

File: src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+            String propertyString = TypeAndPropertyKeyStore.getInstance().
+                mapShortPropertyKeyToString(propertyKey);
+            if (typeKey == TypeAndPropertyKeyStore.ANY) {
+                String exceptionMessage = "The index on property " + propertyString + " already " +
+                    "exists on all types.";
+                throw new IndexAlreadyExistsException(exceptionMessage);
+            } else {
+                String typeString = TypeAndPropertyKeyStore.getInstance().
+                    mapShortToStringType(typeKey);
+                String exceptionMessage = "The index on property " + propertyString + " already " +
```
Comment (semihsalihoglu-uw) [inline src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java]: Ditto.

File: src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java [hunk 8]
```diff
@@ -0,0 +1,396 @@
...
+     * @param currentIndexedTypes The set of types that are currently indexed.
+     * @param typeKey The type that is going to be indexed.
+     * @param propertyKey The property that is going to be indexed.
+     */
+    private void validateIndex(Set<Short> currentIndexedTypes, Short typeKey, Short propertyKey) {
+        if (currentIndexedTypes.contains(TypeAndPropertyKeyStore.ANY)) {
+            String propertyString = TypeAndPropertyKeyStore.getInstance().
+                mapShortPropertyKeyToString(propertyKey);
+            if (typeKey == TypeAndPropertyKeyStore.ANY) {
+                String exceptionMessage = "The index on property " + propertyString + " already " +
```
Comment (semihsalihoglu-uw) [inline src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java]: You can inline exception message.

File: src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java [hunk 9]
```diff
@@ -0,0 +1,396 @@
...
+            for (short type = 1; type <= typeCount; type += 1) {
+                Integer typeAndPropertyToIndex = getTypePropertyIndexKey(type, propertyKey);
+                indices.putIfAbsent(typeAndPropertyToIndex, new HashMap<>());
+            }
+        } else {
+            Integer typeAndPropertyToIndex = getTypePropertyIndexKey(typeKey, propertyKey);
+            indices.put(typeAndPropertyToIndex, new HashMap<>());
+        }
+
+        // Populate the new indices.
```
Comment (semihsalihoglu-uw) [inline src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java]: I'd remove this private comment.

File: src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java [hunk 10]
```diff
@@ -0,0 +1,396 @@
...
+        validateIndex(currentIndexedTypes, typeKey, propertyKey);
+
+        // If we're indexing on all types, the indexedPropertiesAndTypes should not maintain a list
+        // of all types in the graph, just store that an index is being maintained on all types.
+        if (TypeAndPropertyKeyStore.ANY == typeKey) {
+            currentIndexedTypes.clear();
+        }
+        currentIndexedTypes.add(typeKey);
+
+        // Create new empty indices
```
Comment (semihsalihoglu-uw) [inline src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java]: I'd remove this private comment.

File: src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java [hunk 11]
```diff
@@ -0,0 +1,396 @@
...
+        }
+
+        // Populate the new indices.
+        for (int vertexId = 0; vertexId <= Graph.getInstance().getHighestVertexId(); vertexId++) {
+            short vertexType = Graph.getInstance().getVertexTypes().get(vertexId);
+            Integer indexKey = getTypePropertyIndexKey(vertexType, propertyKey);
+            if (TypeAndPropertyKeyStore.ANY != typeKey && vertexType != typeKey) {
+                continue;
+            }
+
```
Comment (semihsalihoglu-uw) [inline src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java]: remove blank line.

File: src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java [hunk 12]
```diff
@@ -0,0 +1,396 @@
...
+                "' is of type '" + dataType.toString() + "'. Expected to index on properties of" +
+                "type String only.");
+        }
+    }
+
+    private String[] tokenizePropertyValue(Object propertyValue) {
+        if (null == propertyValue) {
+            return new String[]{};
+        }
+        String value = (String) propertyValue;
```
Comment (semihsalihoglu-uw) [inline src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java]: Inline value.

File: src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java [hunk 13]
```diff
@@ -0,0 +1,396 @@
...
+     * @param property The property that is being queried.
+     * @param value The value of the property to match.
+     * @return The vertices of that type with the specified value in the specified property.
+     */
+    Set<Integer> getVertices(short startVertexType, short property, String value) {
+        if (TypeAndPropertyKeyStore.ANY != startVertexType) {
+            return getVerticesGivenTypeIsNotANY(startVertexType, property, value);
+        }
+
+        // startVertexType is Type.ANY so go through all the types and collect the vertices whose
```
Comment (semihsalihoglu-uw) [inline src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java]: I'd remove private comment. The code is clear enough.

File: src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java [hunk 14]
```diff
@@ -0,0 +1,396 @@
...
+
+    /**
+     * Update a vertex's properties and update the index store appropriately.
+     *
+     * @param vertexId The ID of the vertex to update.
+     * @param vertexType The type of the vertex that is changing its properties.
+     * @param oldVertexProperties The old properties of the vertex to be removed.
+     * @param newVertexProperties The new properties of the vertex to be added.
+     */
+    void updateVertexPropertiesInIndex(int vertexId, short vertexType,
```
Comment (semihsalihoglu-uw) [inline src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java]: I'd rename this to: updateIndexWithNewVertexProperties. You're not updating vertex properties, you're updating the index.

File: src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java [hunk 15]
```diff
@@ -0,0 +1,396 @@
...
+        Map<Short, Object> oldVertexProperties,
+        Map<Short, Pair<DataType, String>> newVertexProperties) {
+        if (null != newVertexProperties) {
```
Comment (semihsalihoglu-uw) [inline src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java]: What is the behaviour you want when newVertexProperties is null? I was expecting when there are no new properties, you just remove the old properties. so I was expecting removeVertexFromIndexedPropertyType(vertexId, vertexType, oldVertexProperties) to be outside the if. That depends on what the meaning of null is here. Whatever the expected behavior is, you should specify it clearly in the comments.

File: src/main/java/ca/waterloo/dsg/graphflow/query/parser/GraphflowVisitor.java [hunk 16]
```diff
@@ -179,6 +179,20 @@ public AbstractStructuredQuery visitCreateVertexPattern(CreateVertexPatternConte
         return structuredQuery;
     }
 
+    @Override
+    public AbstractStructuredQuery visitCreateIndexPattern(CreateIndexPatternContext ctx) {
+        StructuredQuery structuredQuery = new StructuredQuery();
+        structuredQuery.setQueryOperation(QueryOperation.CREATE_INDEX);
+        String type = null;
```
Comment (semihsalihoglu-uw) [inline src/main/java/ca/waterloo/dsg/graphflow/query/parser/GraphflowVisitor.java]: Move String type = null one line below. General rule is that variables are defined as close as possible to where they will be used, which is inside the if below.

File: src/test/java/ca/waterloo/dsg/graphflow/graph/IndexStoreTest.java [hunk 17]
```diff
@@ -0,0 +1,431 @@
...
+        IndexStore.getInstance().serializeAll(serFolder.getRoot().getAbsolutePath());
+        IndexStore.reset();
+
+        IndexStore.getInstance().deserializeAll(serFolder.getRoot().getAbsolutePath());
+        Assert.assertEquals(oldIndecies, IndexStore.getInstance().indices);
+        Assert.assertEquals(oldIndexedPropertiesAndTypes, IndexStore.getInstance().
+            indexedPropertiesAndTypes);
+    }
+
+    // IndexStore Test Helpers
```
Comment (semihsalihoglu-uw) [inline src/test/java/ca/waterloo/dsg/graphflow/graph/IndexStoreTest.java]: Remove comment.

File: src/test/java/ca/waterloo/dsg/graphflow/graph/IndexStoreTest.java [hunk 18]
```diff
@@ -0,0 +1,431 @@
...
+    }
+
+    private void createIndex(String indexString) {
+        String indexCreationQuery = "CREATE INDEX ON " + indexString + ";";
+        StructuredQuery structuredQuery = new StructuredQueryParser().parse(indexCreationQuery);
+        new IndexCreationQueryPlanner(structuredQuery).plan().execute();
+    }
+
+    private void runCreate(String createQuery) {
+        StructuredQuery createStructuredQuery = new StructuredQueryParser().parse(createQuery);
```
Comment (semihsalihoglu-uw) [inline src/test/java/ca/waterloo/dsg/graphflow/graph/IndexStoreTest.java]: inline createStructuredQuery.

File: src/test/java/ca/waterloo/dsg/graphflow/graph/IndexStoreTest.java [hunk 19]
```diff
@@ -0,0 +1,431 @@
...
+        IndexStore.getInstance().createIndex(typeKey, propertyKey);
+    }
+
+    private void createIndex(Short propertyKey) {
+        IndexStore.getInstance().createIndex(TypeAndPropertyKeyStore.ANY, propertyKey);
+    }
+
+    private void createIndex(String indexString) {
+        String indexCreationQuery = "CREATE INDEX ON " + indexString + ";";
+        StructuredQuery structuredQuery = new StructuredQueryParser().parse(indexCreationQuery);
```
Comment (semihsalihoglu-uw) [inline src/test/java/ca/waterloo/dsg/graphflow/graph/IndexStoreTest.java]: Inline structuredQuery for sure and maybe even indexCreationQuery. Your call.

File: src/test/java/ca/waterloo/dsg/graphflow/graph/IndexStoreTest.java [hunk 20]
```diff
@@ -0,0 +1,431 @@
...
+        short nameProperty = 1;
+
+        Assert.assertFalse(IndexStore.getInstance().isIndexed(personType, nameProperty));
+        createIndex(personType, nameProperty);
+        Assert.assertTrue(IndexStore.getInstance().isIndexed(personType, nameProperty));
+        Set<Integer> sids = IndexStore.getInstance().getVertices(personType, nameProperty,
+            "Sid");
+        Assert.assertEquals(sids.size(), 1);
+        Assert.assertTrue(sids.contains(4));
+
```
Comment (semihsalihoglu-uw) [inline src/test/java/ca/waterloo/dsg/graphflow/graph/IndexStoreTest.java]: I'm skimming over the tests. They look good overall. I'd just make sure that these blank lines that you're leaving are justified.

File: src/test/java/ca/waterloo/dsg/graphflow/graph/IndexStoreTest.java [hunk 21]
```diff
@@ -0,0 +1,431 @@
...
+    }
+
+    @Test
+    public void testChangeVertexPropertyWhenIndexedOnASpecificType() {
+        createSimpleGraphWithManyTypes();
+        short personType = 1;
+        short nameProperty = 1;
+
+        createIndex(personType, nameProperty);
+
```
Comment (semihsalihoglu-uw) [inline src/test/java/ca/waterloo/dsg/graphflow/graph/IndexStoreTest.java]: For example here and above, why do you leave a blank line?

File: src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java (hunk 3 above)
Comment (pbardea) [reply src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java]: Make public

File: src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java (hunk 15 above)
Comment (pbardea) [reply src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java]: Do this check at the call-site. Here mention that it's a NOP if nothing happens.

File: src/main/java/ca/waterloo/dsg/graphflow/graph/Graph.java [hunk 22]
```diff
@@ -174,8 +178,18 @@ public void addEdgeTemporarily(int fromVertex, int toVertex, short fromVertexTyp
         }
         setVertexType(fromVertex, fromVertexType);
```
[the remark this answers is not in the document]
Comment (pbardea) [reply src/main/java/ca/waterloo/dsg/graphflow/graph/Graph.java]: Done

File: src/main/java/ca/waterloo/dsg/graphflow/graph/Graph.java (hunk 1 above)
Comment (pbardea) [reply src/main/java/ca/waterloo/dsg/graphflow/graph/Graph.java]: Updated to check if there are oldVeretxProperties and avoid unnecessary exceptions.

File: src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java (hunk 2 above)
Comment (pbardea) [reply src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java]: done

File: src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java (hunk 10 above)
Comment (pbardea) [reply src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java]: done

File: src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java (hunk 9 above)
Comment (pbardea) [reply src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java]: done

File: src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java (hunk 3 above)
Comment (pbardea) [reply src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java]: Done

File: src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java (hunk 13 above)
Comment (pbardea) [reply src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java]: done

File: src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java (hunk 15 above)
Comment (pbardea) [reply src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java]: done

File: src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java (hunk 4 above)
Comment (pbardea) [reply src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java]: done

File: src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java (hunk 8 above)
Comment (pbardea) [reply src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java]: done

File: src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java (hunk 7 above)
Comment (pbardea) [reply src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java]: done

File: src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java (hunk 6 above)
Comment (pbardea) [reply src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java]: done

File: src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java (hunk 5 above)
Comment (pbardea) [reply src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java]: done

File: src/main/java/ca/waterloo/dsg/graphflow/query/parser/GraphflowVisitor.java (hunk 16 above)
Comment (pbardea) [reply src/main/java/ca/waterloo/dsg/graphflow/query/parser/GraphflowVisitor.java]: done

File: src/test/java/ca/waterloo/dsg/graphflow/graph/IndexStoreTest.java (hunk 18 above)
Comment (pbardea) [reply src/test/java/ca/waterloo/dsg/graphflow/graph/IndexStoreTest.java]: done

File: src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java (hunk 14 above)
Comment (pbardea) [reply src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java]: done

File: src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java (hunk 12 above)
Comment (pbardea) [reply src/main/java/ca/waterloo/dsg/graphflow/graph/IndexStore.java]: done

File: src/test/java/ca/waterloo/dsg/graphflow/graph/IndexStoreTest.java (hunk 17 above)
Comment (pbardea) [reply src/test/java/ca/waterloo/dsg/graphflow/graph/IndexStoreTest.java]: done

File: src/main/java/ca/waterloo/dsg/graphflow/query/plan/CreateQueryPlan.java [hunk 23]
```diff
@@ -2,6 +2,7 @@
 
 import ca.waterloo.dsg.graphflow.exception.IncorrectDataTypeException;
 import ca.waterloo.dsg.graphflow.graph.Graph;
+import ca.waterloo.dsg.graphflow.graph.IndexStore;
```
Comment (nare469) [inline src/main/java/ca/waterloo/dsg/graphflow/query/plan/CreateQueryPlan.java]: unused import

**Comment by sachdevs** [comment]
LGTM 💸

**Pull request merged**